docs/design_notes/module_lifecycle.rst

```diff
@@ -0,0 +1,378 @@
+The Module lifecycle
+######################
+
+.. testsetup::
+
+  from typing import Any, Callable, Iterable
+  import flax
+  from flax import linen as nn
+  from jax import random
+  import jax
+
+
+This design note is intended for users who are already familiar with linen Modules but want to understand more about the design principles behind the abstraction. This note should give you a good understanding of the assumptions and guarantees the Module API is built upon. If you have no practical experience with Modules yet, check out the `MNIST Tutorial <https://flax.readthedocs.io/en/latest/notebooks/annotated_mnist.html>`_.
+
+linen Modules offer a pythonic abstracton on top of Flax core. The ``Module`` abstraction allows you to create classes that have state, parameters and randomness on top of JAX. This is a practical guide to the design and behavior of the ``Module`` class. By the end, you should feel comfortable to go off the beaten track and use Modules in new ways.
+
+
+Overview
+***********
+
+Definition
+=============
+
+Let's start with a high-level overview of the Module lifecycle. We start by defining a simple Module:
```

Comment:
Nit: to avoid repeating "start", maybe try

```suggestion
Let's start with a high-level overview of the Module lifecycle. First, define a simple Module:
```